src/main/java/io/cryostat/discovery/DiscoveryStorage.java

```diff
@@ -318,41 +319,98 @@ private List<AbstractNode> modifyChildrenWithJvmIds(
 
-        if (notify) {
-            Set<TargetNode> previousLeaves = findLeavesFrom(original);
-            Set<TargetNode> currentLeaves = findLeavesFrom(currentTree);
+        Set<TargetNode> previousLeaves = findLeavesFrom(originalTree);
+        Set<TargetNode> currentLeaves = findLeavesFrom(currentTree);
 
-            Set<TargetNode> added = new HashSet<>(currentLeaves);
-            added.removeAll(previousLeaves);
+        Set<TargetNode> updated = getUpdatedLeaves(previousLeaves, currentLeaves);
+        updated.stream()
+                .map(TargetNode::getTarget)
+                .forEach(sr -> notifyAsyncTargetDiscovery(EventKind.MODIFIED, sr));
 
-            Set<TargetNode> removed = new HashSet<>(previousLeaves);
-            removed.removeAll(currentLeaves);
+        Set<TargetNode> added =
+                removeAllUpdatedLeaves(
+                        getAddedOrUpdatedLeaves(previousLeaves, currentLeaves), updated);
+        added.stream()
+                .map(TargetNode::getTarget)
+                .forEach(sr -> notifyAsyncTargetDiscovery(EventKind.FOUND, sr));
+
+        Set<TargetNode> removed =
+                removeAllUpdatedLeaves(
+                        getRemovedOrUpdatedLeaves(previousLeaves, currentLeaves), updated);
+        removed.stream()
+                .map(TargetNode::getTarget)
+                .forEach(sr -> notifyAsyncTargetDiscovery(EventKind.LOST, sr));
 
-            removed.stream()
-                    .map(TargetNode::getTarget)
-                    .forEach(sr -> notifyAsyncTargetDiscovery(EventKind.LOST, sr));
-            added.stream()
-                    .map(TargetNode::getTarget)
-                    .forEach(sr -> notifyAsyncTargetDiscovery(EventKind.FOUND, sr));
-        }
         return currentTree.getChildren();
     }
 
+    public Set<TargetNode> getAddedOrUpdatedLeaves(
+            Set<TargetNode> previousLeaves, Set<TargetNode> currentLeaves) {
+        Set<TargetNode> added = new HashSet<>(currentLeaves);
+        added.removeAll(previousLeaves);
+        return added;
+    }
+
+    public Set<TargetNode> getRemovedOrUpdatedLeaves(
+            Set<TargetNode> previousLeaves, Set<TargetNode> currentLeaves) {
+        Set<TargetNode> removed = new HashSet<>(previousLeaves);
+        removed.removeAll(currentLeaves);
+        return removed;
+    }
+
+    public Set<TargetNode> getUpdatedLeaves(
+            Set<TargetNode> previousLeaves, Set<TargetNode> currentLeaves) {
+        Set<TargetNode> added = getAddedOrUpdatedLeaves(previousLeaves, currentLeaves);
+        Set<TargetNode> removed = getRemovedOrUpdatedLeaves(previousLeaves, currentLeaves);
+        Set<TargetNode> updated = new HashSet<>();
+
+        // Manual set intersection since ServiceRef also compares jvmId
+        for (TargetNode addedTNode : added) {
+            ServiceRef atServiceRef = addedTNode.getTarget();
+            for (TargetNode removedTNode : removed) {
+                if (new EqualsBuilder()
```

Comment:
This could just be `Objects.equals(atServiceRef.getServiceUri(), removedTNode.getTarget().getServiceUri())`.

Reply:
This is enforcing that a target is modified and two TargetNodes are considered a modification if the ServiceRef serviceUri is the same. I'm wondering does this make sense and is enough? e.g. Do other ServiceRef attributes(not jvmId) need to be compared as well?

Reply:
I am thinking maybe later we allow editting other attributes, for example, alias or labels (not sure how tho)?

Reply:
^ I don't think so, those are provided by discovery plugins (even builtins) and I don't want that to be a two-way communication between the plugin and Cryostat to coordinate overriding attributes like labels. Two-way communication to establish and maintain the plugin registration is enough, the rest should just be a one-way data flow from plugin to Cryostat. The other way to do it would be for Cryostat to store these modifications and apply them as overlays atop of whatever data the plugins have pushed, but I'd rather avoid that too. This also opens a whole can of worms for rule activation, credentials matching, etc. that use match expressions and where modifications to targets should cascade into re-evaluations of any resources that have match expressions.

Reply:
Ahh right I remember you explained this that issue too. Seems to cause some trouble. Do you also want to compare other attributes too (except jvmId)?

Reply:
The targets' URLs should already be unique and distinct from each other on any given update of a plugin's subtree - if there are two leaf nodes with identical URLs then that's a discovery plugin bug. So for the purpose of detecting added/changed/removed elements in the set of leaf nodes, the URL can be considered as a unique ID. It may happen for some discovery plugins, corresponding to some application platforms, that target properties like labels or alias are actually configurable or otherwise runtime dynamic (but this would be implemented by the plugin itself, not through any interface on Cryostat), and so these things are not part of the target's identity.

As I'm saying this I realize that this means that discovery mechanisms like JDP which have an intrinsic lag time between an application going offline and the registration that this has occurred will have to be careful about how they publish discovery subtrees. If the target reappears with the same URL before the discovery mechanism has detected its prior disappearance, then a naive discovery implementation may cause this to register as a `MODIFIED` event followed by a `LOST` event.

Reply:
>  If the target reappears with the same URL before the discovery mechanism has detected its prior disappearance, then a naive discovery implementation may cause this to register as a MODIFIED event followed by a LOST event.

Oh this problem is pretty severe. Is this something Cryostat is encoutering or already implemented to safeguard against?

Reply:
Well, the original safeguard against this class of problem was the `LOST/FOUND` event pair emitted by the original discovery mechanisms. Now that `MODIFIED` is an option, the discovery mechanisms probably need to handle it in some more intelligent way if this is a possibility for them. I think it's probably only a likely problem for JDP right now.

Reply:
While we're here, the other built-in discovery mechanisms like `KubeApiPlatformClient` also need updating to emit `MODIFIED` events instead of the `LOST/FOUND` pair.

Reply:
Ahh make sense, thanks :D I refactored and added MOFIDIFED for `KubeApiPlatformClient` now. Stuck in test but forgot to reset latch :D